src/coreclr/jit/morph.cpp

```diff
@@ -7047,19 +7007,46 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee, const char** failReason)
     // as we need to keep our frame around.
     if (fgCallHasMustCopyByrefParameter(callee))
     {
-        reportFastTailCallDecision("Callee has a byref parameter");
+        *failReason = "Callee has a byref parameter";
         return false;
     }
 
-    reportFastTailCallDecision(nullptr);
+    *failReason = nullptr;
     return true;
 #else // FEATURE_FASTTAILCALL
-    if (failReason)
-        *failReason = "Fast tailcalls are not supported on this platform";
+    *failReason = "Fast tailcalls are not supported on this platform";
     return false;
 #endif
 }
 
+//------------------------------------------------------------------------
+// fgShouldFastTailCall: Do profitability checks for whether to do a fast tailcall.
+//
+// Arguments:
+//    call - The call to check
+//    failReason - If this method returns false, the reason why. Can be nullptr.
+//
+// Return Value:
+//    Returns true or false based on whether the callee can be fast tailcalled.
+//
+bool Compiler::fgShouldFastTailCall(GenTreeCall* call, const char** failReason)
+{
+#ifdef TARGET_AMD64
+    // Indirected R2R calls on x64 need indirection cell that is normally
```

Comment:
Tailcalls have often secondary effect like reducing the size of the stackframe.

Do you have some aggregate statistics that disabling tailcalls in R2R is the right trade-off? (My intuition would be that it is better to do the tailcalls on average despite the indirection overhead.)

Reply:
I don't have that data, no. It sounds good to me to always do it anyway to keep the test matrix smaller (otherwise we will have very limited coverage of tail. prefixed direct calls in R2R).

There is one additional 'invisible' source of overhead that I'm not too sure how to remove: currently for tailcalls through indirection we end up creating two different delay load helper stubs in the R2R image on x64, one that gets it from caller instructions and one that gets it through register (when we realize later that we will do a fast tailcall). The problem is that we create the entry point very eagerly in crossgen2 and we don't really have a way of backing out. Do you have any ideas for this? Maybe some kind of ref-count mechanism in CG2?